### PR TITLE
Feature/cplp 1235 add single user ownidp

### PR DESCRIPTION
**Pull request by ntruchsess**
implements:
https://jira.catena-x.net/browse/CPLP-1235 (API: POST add single user ownIdP)
https://jira.catena-x.net/browse/CPLP-1227 (API: POST User Bulk Upload SharedIdP (with support of csv))

### CODE REVIEW

**Review by Phil91 — CHANGES_REQUESTED** [review]

File: src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/UserController.cs [hunk 1]
```diff
@@ -28,327 +28,374 @@
...
+    public Task<int> DeleteOwnUser([FromRoute] Guid companyUserId) =>
+        this.WithIamUserId(iamUserId => _logic.DeleteOwnUserAsync(companyUserId, iamUserId));
 
-        [Obsolete]
-        [HttpPut]
-        [Authorize(Roles = "modify_user_account")]
-        [Route("bpn")]
-        public Task BpnAttributeAdding([FromBody] IEnumerable<UserUpdateBpn> usersToAddBpn) =>
-            _logic.AddBpnAttributeAsync(usersToAddBpn);
+    [Obsolete]
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/UserController.cs:317]: I don't know whether this is still being called by the frontend, we need to ask @jjeroch of create a bug-ticket for that. Here it is out of scope.

File: src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/UserController.cs [hunk 2]
```diff
@@ -28,327 +28,374 @@
...
+    /// Example: POST: api/administration/users/identityprovider/owncompany/identityprovider/{identityProviderId}/users
+    /// </remarks>
+    /// <response code="200">Returns a file of users.</response>
+    /// <response code="400">user is not associated with a company.</response>
+    /// <response code="415">Content type didn't match the expected value.</response>
+    /// <response code="502">Bad Gateway Service Error.</response>
+    [HttpPost]
+    [Authorize(Roles = "add_user_account")]
+    [Route("owncompany/identityprovider/{identityProviderId}/users")]
+    public async Task<CreatedAtRouteResult> CreateOwnIdpOwnCompanyUser([FromBody] UserCreationInfoIdp userToCreate, [FromRoute] Guid identityProviderId)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/CatenaX.NetworkServices.Administration.Service/Controllers/UserController.cs:91]: done

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/CatenaX.NetworkServices.Provisioning.Library.csproj [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
+  <ItemGroup>
+    <PackageReference Include="PasswordGenerator" Version="2.1.0" />
+    <PackageReference Include="System.Linq.Async" Version="6.0.1" />
+  </ItemGroup>
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+</Project>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/CatenaX.NetworkServices.Provisioning.Library.csproj]: done

File: src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/CatenaX.NetworkServices.PortalBackend.DBAccess.csproj [hunk 4]
```diff
@@ -1,20 +1,16 @@
...
+  <ItemGroup>
+    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.4" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.3.1" />
+  </ItemGroup>
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+</Project>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/portalbackend/CatenaX.NetworkServices.PortalBackend.DBAccess/CatenaX.NetworkServices.PortalBackend.DBAccess.csproj]: done

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+                        organisationName: companyName,
+                        businessPartnerNumber: businessPartnerNumber
+                    )
+                ).ConfigureAwait(false);
+
+                userRepository.CreateIamUser(companyUser, centralUserId);
+
+                await _provisioningManager.AddProviderUserLinkToCentralUserAsync(centralUserId, new IdentityProviderLink(alias, providerUserId, user.UserName)).ConfigureAwait(false);
+
+                await AssignRolesToNewUser(userRolesRepository, user.Roles, companyUser.Id, centralUserId, clientId, userRoleIds);
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs:105]: done

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+        if (result.CompanyName == null)
+        {
+            throw new ConflictException($"assertion failed: companyName of company {result.CompanyId} should never be null here");
+        }
+
+        return new CompanyNameIdpAliasData(result.CompanyId, result.CompanyName, result.BusinessPartnerNumber, result.companyUserId, result.IdpAlias, result.IsSharedIdp);
+    }
+
+    private async Task ValidateDuplicateUsers(IUserRepository userRepository, string alias, UserCreationInfoIdp user, Guid companyId)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs:143]: done

File: src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs [hunk 7]
```diff
@@ -29,588 +30,737 @@
...
-            return await _portalRepositories.SaveAsync().ConfigureAwait(false);
         }
+        await foreach (var roles in appRepository.GetClientRolesAsync(appId, languageShortName).ConfigureAwait(false))
+        {
+            yield return new ClientRoles(roles.RoleId, roles.Role, roles.Description);
+        }
+    }
 
-        public async IAsyncEnumerable<Guid> DeleteOwnCompanyUsersAsync(IEnumerable<Guid> companyUserIds, string adminUserId)
+    public async Task<CompanyUserDetails> GetOwnCompanyUserDetails(Guid companyUserId, string adminUserId)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/administration/CatenaX.NetworkServices.Administration.Service/BusinessLogic/UserBusinessLogic.cs:418]: done

File: src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+                }
+            }
+            catch(KeycloakEntityNotFoundException)
+            {
+                // when searching for duplicates this is not a validation-error
+            }
+        }
+    }
+
+    private async Task AssignRolesToNewUser(IUserRolesRepository userRolesRepository, IEnumerable<string> roles, Guid companyUserId, string centralUserId, string clientId, IDictionary<string,Guid> userRoleIds)
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/provisioning/CatenaX.NetworkServices.Provisioning.Library/Service/UserProvisioningService.cs:162]: done

File: src/framework/CatenaX.NetworkServices.Framework.IO/CatenaX.NetworkServices.Framework.IO.csproj [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+</Project>
```
[the remark this answers is not in the document]
Comment (ntruchsess) [reply src/framework/CatenaX.NetworkServices.Framework.IO/CatenaX.NetworkServices.Framework.IO.csproj]: done

**Review by Phil91 — APPROVED** [review]

**Pull request merged**